examples/NFD/stateful_firewall/stateful_firewall.cpp

```diff
@@ -0,0 +1,317 @@
+* Details:    This code is the implementation of state_firewall from NFD project,
+              a C++ based NF developing framework designed by Wenfei's group
+              from IIIS, Tsinghua University, China.
+*************************************************************************************/
+
+#include <errno.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <unistd.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <rte_common.h>
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+
+#include "onvm_nflib.h"
+#include "onvm_pkt_helper.h"
+
+#ifdef __cplusplus
+}
+#endif
+
+// NFD ADD
+#include <pcap.h>
+#include <signal.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "basic_classes.h"
+#include "decode.h"
+
+using namespace std;
+
+#define NF_TAG "stateful_firewall"
+
+/* Struct that contains information about this NF */
+struct onvm_nf_info *nf_info;
+
+/* number of package between each print */
+static uint32_t print_delay = 1000000;
+
+static uint32_t destination;
+
+/*******************************NFD features********************************/
+Flow::Flow(u_char *packet, int totallength) {
+        /*Decoding*/
+        this->pkt = packet;
+        int ethernet_header_length = 14;
+        /* Header lengths in bytes */
+        EtherHdr *e_hdr = (EtherHdr *)packet;
+        if (ntohs(e_hdr->ether_type) == 0x8100)
+                ethernet_header_length = 14 + 4; /* For 802.1Q Virtual LAN */
+        else
+                ethernet_header_length = 14; /* For general wired */
+
+        IPHdr *ip_hdr = (IPHdr *)(packet + ethernet_header_length);
+        int src_addr = ntohl(ip_hdr->ip_src.s_addr);
+        this->headers[Sip] = new IP(src_addr, 32);
+        int dst_addr = ntohl(ip_hdr->ip_dst.s_addr);
+        this->headers[Dip] = new IP(dst_addr, 32);
+}
+void
+Flow::clean() {
+        /*Encoding*/
+}
+
+long int _counter = 0;
+long int _drop = 0;
+unordered_map<string, int> F_Type::MAP = unordered_map<string, int>();
+unordered_map<string, int> F_Type::MAP2 = unordered_map<string, int>();
+int
+process(Flow &f);
+Flow f_glb;
+
+/////time////
+struct timeval begin_time;
+struct timeval end_time;
+
+void
+_init_() {
+        (new F_Type())->init();
+}
+
+// this setting was set by the model.txt
+IP _t1("192.168.22.0/24");
+
+State<unordered_set<IP>> seen(*(new unordered_set<IP>()));
+
+int
+process(Flow &f) {
+        if (*((IP *)f.headers[Sip]) <= _t1) {
+                seen[f] = union_set<unordered_set<IP>>(
+                    seen[f], create_set<IP>(*(new unordered_set<IP>()), 1, &((*(IP *)f.headers[Dip]))));
+        } else if ((*((IP *)f.headers[Sip]) != _t1) && (seen[f].find((*(IP *)f.headers[Sip])) != seen[f].end())) {
```

Comment:
Is this line extra?

Reply:
the empty branch means that we do no operation on the packets but just pass it.